text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,128 @@
+
+**Never create a dependency conflict between packages which rely on different versions of OpenTelemetry. Avoid breaking all stable public APIs.**  
+Backwards compatibility is a strict requirement. Instrumentation APIs cannot create a version conflict, ever. Otherwise, OpenTelemetry cannot be embedded in widely shared libraries, such as web frameworks. Theoretically, APIs can be deprecated and eventually removed, but this is a process measured in years.
+
+**Allow for multiple levels of package stability within the same release.**  
+Provide maintainers a clear process for developing new, experimental APIs alongside stable APIs. DIfferent packages within the same release will have different levels of stability.
+
+## Relevant architecure
+![drawing](img/0143_cross_cutting.png)
+
+At the highest architectural level, OpenTelemetry is organized into signals. Each signal provides a specialized form of observability. For example, tracing, metrics, and baggage are three seperate signals. Signals share a common subsystem – context propagation – but they function independently from each other.
+
+Each signal provides a mechanism for software to describe itself. A codebase, such as an API handler or a database client, takes a dependency on various signals in order to describe itself. OpenTelemetry instrumentation code is then mixed into the other code within that codebase. This makes OpenTelemetry a **cross-cutting concern** - a piece of software which must be mixed into many other pieces of software in order to provide value. Cross-cutting concerns, by their very nature, violate a core design principlie – seperation of concerns. As a result, OpenTelemetry requires extra care and attention to avoid creating issues for the codebases which depend upon it for instrumentaion.
+
+OpenTelemetry is designed to seprate the portion of each signal which must be imported as cross-cutting concerns from the portions of OpenTelemetry which can be managed independently. OpenTelemetry is also designed to be a pluggable framework. To accomplish this these goals, each signal consists of four types of packages:
+
+**API -** API packages consist of the public interfaces used for instrumentation. Any portion of OpenTelemetry which 3rd-party libraries and application code depend upon. To manage different levels of stability, every signal has its own, independent API package. These individual APIs may also be bundled up into a shared global API, for convenience.
+
+**SDK -** The implemention of the API. The SDK is managed by the application owner. Note that the SDKs includes additional public interfaces which are not considered part of the API package, as they are not cross-cutting concerns. These public interfaces include constructors, configuration interfaces, and plugin interfaces.
```

Comment:
Here we should better distinguish between application owner and library developer. The latter should not touch an SDK, it is only for the former.

Reply:
Absolutely agree, @iNikem. In addition, we need to make sure that the SDK specification clearly lays out what the supported extension points/supported SDK "APIs" must be for every SDK implementation. 

There is a difference between extending an SDK, and configuring an SDK. These should be two classes of SDK APIs that we need to consider how we're going to manage from a versioning/release process.

Off the top of my head, the following is the list of what probably needs to be supported with strict ABI as potential extension points/apis in the SDK: Exporters, SpanProcessors, IdGenerators, Resource Semantic Attributes, Exportable SpanData, ReadableSpan, ReadWriteSpan, Sampler [I'm sure I'm missing a few].

Reply:
I totally agree with this sentiment, but the wording is a little difficult. Specifically, how do we differentiate between application business logic which should only touch the API, and the kind of SDK management subsystems which the application owner may write? Both types of code are technically "application code" written by the "application owner."

Reply:
Advanced Samplers may be written and supported by someone other than the Application owner, and pulled in as a dependency at startup time, though.